apps/openmw/mwsound/soundmanagerimp.cpp

```diff
@@ -367,23 +366,23 @@ namespace MWSound
             }
 
             mMusicFiles[mCurrentPlaylist] = filelist;
-
+            for(int it = 0; it < filelist.size(); it++)
+                mLastPlayedMusic.push_back(it);
         }
         else
             filelist = mMusicFiles[mCurrentPlaylist];
 
         if(filelist.empty())
             return;
 
-        int i = Misc::Rng::rollDice(filelist.size());
-
-        // Don't play the same music track twice in a row
-        if (filelist[i] == mLastPlayedMusic)
-        {
-            i = (i+1) % filelist.size();
-        }
+        //Do a Fisher-Yates shuffle
+        if(mMusicFiles.size() == 0)
+            for(int it = 0; it < filelist.size(); it++)
+                mLastPlayedMusic.push_back(it);
 
-        advanceMusic(filelist[i]);
+        int i = Misc::Rng::rollDice(mLastPlayedMusic.size());
+        advanceMusic(filelist[mLastPlayedMusic[i]]);
+        mLastPlayedMusic.erase(mLastPlayedMusic.begin()+i-1);
```

Comment:
This will crash if ```i == 0``` (if the RNG happens to roll 0, or ```mLastPlayedMusic```/```mMusicFiles``` are 0 size. I don't think that ```-1```should be there. Also, ```mLastPlayedMusic``` seems to be a misnomer since it's a list of indices for playlist entries that haven't yet played.